src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Roslyn.Utilities;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract partial class SymbolKey
+    {
+        private class TupleTypeSymbolKey : AbstractSymbolKey<TupleTypeSymbolKey>
+        {
+            private readonly SymbolKey[] _types;
+            private readonly string[] _names;
+
+            internal TupleTypeSymbolKey(ITupleTypeSymbol symbol, Visitor visitor)
+            {
+                _types = symbol.Types.Select(t => GetOrCreate(t, visitor)).ToArray();
+
+                ImmutableArray<string> names = symbol.Names;
+                _names = names.IsDefault ? null : names.ToArray();
+            }
+
+            public override SymbolKeyResolution Resolve(Compilation compilation, bool ignoreAssemblyKey, CancellationToken cancellationToken)
+            {
+                IEnumerable<ITupleTypeSymbol> tupleType = Resolve(compilation, ignoreAssemblyKey);
+                return CreateSymbolInfo(tupleType);
+            }
+
+            private IEnumerable<ITupleTypeSymbol> Resolve(
+                Compilation compilation,
+                bool ignoreAssemblyKey)
+            {
+                ITypeSymbol[] types = _types.Select(a => a.Resolve(compilation, ignoreAssemblyKey).Symbol as ITypeSymbol).ToArray();
+                if (types.Any(a => a == null))
+                {
+                    yield break;
+                }
+
+                yield return compilation.CreateTupleTypeSymbol(types, _names);
+            }
+
+            internal override bool Equals(TupleTypeSymbolKey other, ComparisonOptions options)
+            {
+                var comparer = SymbolKeyComparer.GetComparer(options);
+
+                return SequenceEquals(other._types, _types, comparer)
+                       && ((other._names == null && _names == null)
+                           || SequenceEquals(other._names, _names, CaseInsensitiveComparison.Comparer));
```

Comment:
Oops. I had checked but misread the code. The case of two nulls is covered in the first check...